lang/src/org/partiql/lang/eval/ExprValueExtensions.kt

```diff
@@ -286,350 +249,25 @@ fun ExprValue.cast(
-                                    )
-                                )
-                            }
-                            type.isText -> try {
-                                // validate that the time string follows the format HH:MM:SS[.ddddd...][+|-HH:MM]
-                                val matcher = genericTimeRegex.toPattern().matcher(stringValue())
-                                if (!matcher.find()) {
-                                    castFailedErr(
-                                        "Can't convert string value to TIME. Expected valid time string " +
-                                            "and the time to be of the format HH:MM:SS[.ddddd...][+|-HH:MM]",
-                                        internal = false
-                                    )
-                                }
-
-                                val localTime = LocalTime.parse(stringValue(), DateTimeFormatter.ISO_TIME)
-
-                                // Note that the [genericTimeRegex] has a group to extract the zone offset.
-                                val zoneOffsetString = matcher.group(2)
-                                val zoneOffset = zoneOffsetString?.let { ZoneOffset.of(it) } ?: defaultTimezoneOffset
-
-                                return valueFactory.newTime(
-                                    Time.of(
-                                        localTime,
-                                        precision ?: getPrecisionFromTimeString(stringValue()),
-                                        when (targetType.scalarType.withTimeZone) {
-                                            true -> zoneOffset
-                                            else -> null
-                                        }
-                                    )
-                                )
-                            } catch (e: DateTimeParseException) {
-                                castFailedErr(
-                                    "Can't convert string value to TIME. Expected valid time string " +
-                                        "and the time format to be HH:MM:SS[.ddddd...][+|-HH:MM]",
-                                    internal = false, cause = e
-                                )
-                            }
-                        }
-                    }
-                    else -> error("Unsupported type: ${targetType.scalarType}")
-                }
-                is ListType -> if (type.isSequence) return valueFactory.newList(asSequence())
-                is SexpType -> if (type.isSequence) return valueFactory.newSexp(asSequence())
-                is BagType -> if (type.isSequence) return valueFactory.newBag(asSequence())
-                // no support for anything else
-                else -> {}
-            }
+    when (targetType) {
+        is StaticScalarType -> scalarTypeSystem.invokeCastOp(this, targetType.toCompileTimeType(), locationMeta).let {
```

Comment:
Here is what I expect should be happening here: 
 * Prerequisite: `ExprValue` has been refactored, similar to what you have done with `StaticType`, so that specific scalar classes like `DecimalExprValue` are now in the plugin and there is a subclass of `ExprValue` - say, `ScalarExprValue`, that is an "envelope" for their instances -- similar in role to `StaticScalarType`.
 * Then, here, `this` (which is an `ExprValue`) is to be analyzed to see if it is a `ScalarExprValue` and if so, its scalar value is extracted and *then* a call is made to a function in a plugin, with this extracted scalar value and the scalar type similarly extracted from `targetType`. 

Right now, this does make a call to the plugin, `scalarTypeSystem.invokeCastOp`, but this call does the extraction, which is too much for the plugin to handle, and it manifests in the wrong references (to core PartiQL) appearing in the plugin's interfaces -- which I complain about there. 

I recall you saying something like "we will pull scalar ExprValue into the plugin later", but it is hard for me to imagine what other meaningful refactoring work can proceed without this ExprValue work being done first. It could even be that some of the complexity in the current plugin interfaces is needlessly caused by holding this ExprValue work back.

Reply:
My current opinion is that, like what you said, we need decouple scalar run-time values from `ExprValue` and decouple scalar run-time types from `ExprValueType`, and migrate them into the OTS interfaces package (probably called `ScalarExprValue` & `ScalarExprValueType`). And ideally, this block code should check if the value is a scalar value. If it is, this block of code transforms it into `ScalarExprValue` first, and then, makes a call to the `CAST` invoke defined in the standard plugin.

Reply:
Does this sound like a feasible solution?

Reply:
Yes, we both write pretty much the same thing here.

My point is that it is logical to do this work ahead of other things.  Continuing within this PR would make more sense to me than merging it and then doing another one, because the plugin interfaces should greatly improve and seeing that within one change set will be more understandable and satisfying overall.